### PR TITLE
Create sonarqube.md

### PR DESCRIPTION
**Pull request by andyblundell**

### CODE REVIEW

File: tools/sonarqube.md [hunk 1]
```diff
@@ -0,0 +1,73 @@
...
+
+## The suggested fix is nonsense!
+SonarQube does two jobs: finds real issues; and raises "smells".
+Smells only give you _suggested fixes_. You do not have to follow its suggestions. But you do have to make the build pass again!
+For example: SonarQube raises 10 smells in a piece of code: tiny, annoying things. It isn't telling you to fix them. You could. And get the build passing again like that - but that's not the point.
+What you - as the expert - need to do is look at the bigger picture. Why is SonarQube raising these? Does this entire class need rewriting? Does it interact with other classes properly? Do we need a properties file instead of hardcoding values? Do we need to look at error handling across the whole system? Is our code written well for reuse and testability?
+What's _really_ causing those underlying issues?
+
+The idea of SonarQube isn't to tickle a few lines of code just to get the build passing! If it's shouting about a lot of small, related issues, it probably means there’s something much bigger that you need to address!
```
Comment (andyblundell) [inline tools/sonarqube.md:70]: Suggested rephrasing:

The idea of SonarQube isn't to unquestioningly implement every individual suggested code change, just to get the build to pass. Instead, we should use these suggestions, especially where we have a concentration of suggestions in one area of the code, to consider if there is something bigger that we need to address to improve our code.

?

File: tools/sonarqube.md [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+
+## Why these rules?
+Everyone has different standards. Agreeing is impossible. We need to impose <something>. If we impose the default SonarQube rules, it will breed familiarity across individuals, teams, even companies.
```
Comment (andyblundell) [inline tools/sonarqube.md:73]: We need to soften this until we have some COP consensus - maybe change to the default rules are establish industry standards , crafted by experts, and a really good place to start? (Or something like that?)

File: quality-checks.md [hunk 3]
```diff
@@ -46,16 +46,16 @@ We recommend tracking progress on an Engineering Quality dashboard, for example:
...
...                                                                                                                   | The higher the code coverage, the more thorough the testing, and therefore the higher the likelihood of detecting functional issues early                                                                                                                  | Must use SonarQube's default rules, profiles and gateways <br/><br/> Build pipeline must fail if gateway test coverage is not met <br/><br/> For new code (any class that was changed as part of a commit), must need coverage specified in default Sonarqube gateway. <br/> For legacy code, coverage amount can be initially lowered to deal with historic tech debt. But must have a plan for increasing coverage over time.                                                                                                                                                       |  [SonarQube](tools/sonarqube.md) (in conjunction with testing frameworks)                                                                            |                                                                                                               |
```
Comment (andyblundell) [inline quality-checks.md:56]: We need that definition of "new" code somewhere - possibly in the sonarqube.md file but with a link from here? (Also applies to the other sonar-type entries in this list)

Comment (striggy77) [reply quality-checks.md:56]: > We need that definition of "new" code somewhere - possibly in the sonarqube.md file but with a link from here? (Also applies to the other sonar-type entries in this list)

Also the "out of the box" for code coverage in SonarQube is 80, whereas we are quoting 70% for green for new code. Suggest these figures should be the same.

File: quality-checks.md [hunk 4]
```diff
@@ -46,16 +46,16 @@ We recommend tracking progress on an Engineering Quality dashboard, for example:
...
+| Security testing       | Security        | Contextual    | Check for security issues (for example injection weaknesses)                                                                                                                                                    | More thorough than security code scanning, but much slower to execute, so both are important to achieve both rapid and thorough security testing | Must use SonarQube's default rules, profiles and gateways <br/><br/> Build pipeline must fail if gateway is not met | [SonarQube](tools/sonarqube.md) (Developer Edition or higher - includes advanced OWASP scanning)|                                                                                                               |
```
[the remark this answers is not in the document]
Comment (dlavender4) [reply quality-checks.md:49]: sorry, yeah, that should have been in the box above that.  will fix and reword a bit

File: tools/sonarqube.md (hunk 1 above)
Comment (dlavender4) [reply tools/sonarqube.md:70]: haha much better - thanks. that original was copied from my rant in Teams :)

File: quality-checks.md (hunk 3 above)
Comment (dlavender4) [reply quality-checks.md:56]: have added link in.  and de-confused the 70% phrasing in sonarqube.md: it was an example of a possible rule - but looked like that was the actual rule we were advocating.

File: tools/sonarqube.md [hunk 5]
```diff
@@ -0,0 +1,76 @@
...
+- Security Hotspots – SonarQube highlights security-sensitive pieces of code that need to be reviewed. Upon review, you'll either find there is no threat or you need to apply a fix to secure the code.
+- Raises "smells". It knows _something_ looks odd, and suggests a fix (only suggests!). Your team can then look at the bigger picture around that code, and fix the smells in whichever way is best.
+
+# Profile
+Quality Profiles are a core component of SonarQube, since they are where you define sets of Rules that - when violated - should raise issues on your codebase (example: Methods should not have a Cognitive Complexity higher than 15). Quality Profiles are defined for individual languages.
+
+While it's recommended to have as few Quality Profiles as possible to ensure consistency across projects, you can define as many Quality Profiles as are necessary to fit your specific needs.
+
+The "Sonar Way" Quality Profiles are a good starting-point as you begin analyzing code, and they start out as the default Quality Profiles for each language, created by knowledge experts.
```
Comment (andyblundell) [inline tools/sonarqube.md:19]: "analysing" :)

File: tools/sonarqube.md [hunk 6]
```diff
@@ -0,0 +1,76 @@
+# Sonarqube
+SonarQube® is an automatic code review tool to detect bugs, vulnerabilities, and code smells in your code. It can integrate with your existing workflow to enable continuous code inspection across your project branches and pull requests.
+
+## Benefits of SonarQube
+SonarQube gives you the tools you need to write clean and safe code:
+
+- SonarLint – SonarLint is a companion product that works in your editor giving immediate feedback so you can catch and fix issues before they get to the repository.
+- Quality Gate – The Quality Gate lets you know if your project is ready for production.
+- Clean as You Code – Clean as You Code is an approach to code quality that eliminates a lot of the challenges that come with traditional approaches. As a developer, you focus on maintaining high standards and taking responsibility specifically in the New Code you're working on.
```
Comment (andyblundell) [inline tools/sonarqube.md:9]: Make "New Code" a link to that section in this page?

File: tools/sonarqube.md [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+- Bug (Reliability domain)
+- Vulnerability (Security domain)
+- Security Hotspot (Security domain)
+
+# Quality Gate
+Quality Gates enforce a quality policy in your organisation by answering one question: is my project ready for release?
+
+To answer this question, you define a set of conditions against which projects are measured. For example "No new blocker issues", "Code coverage on new code greater than X%".
+
+You can define gates for "new code" and for all code.
```
Comment (andyblundell) [inline tools/sonarqube.md:34]: Maybe toughen "you can" to "we recommend that you"?

File: tools/sonarqube.md [hunk 8]
```diff
@@ -0,0 +1,76 @@
...
+This approach facilitates getting new code right, without having to instantly fix legacy code.
```
Comment (andyblundell) [inline tools/sonarqube.md:35]: The approach of just defining rules for new code? Shall we say we recommend that you ideally should define rules for all code, but failing that you should *at least* have rules for new code, which facilitates <<the words above>> ??

File: tools/sonarqube.md [hunk 9]
```diff
@@ -0,0 +1,76 @@
...
+
+The recommendation is that we use the Sonarqube Way for all projects. Teams will eventually diverge, but these diversions need to be closely monitored by the Tech Lead and the team.
+
+# Configuring CI pipeline
+Making SonarQube part of a Continuous Integration process is recommended: a build should fail if the code analysis did not satisfy the Quality Gate condition (see [Quality Checks](../quality-checks.md))
+
+# FAQs
+
+## SonarQube is going to add weeks to my workload!
+Initially, yes, it might. Especially if you're introducing it to an existing, legacy codebase.
```
Comment (andyblundell) [inline tools/sonarqube.md:56]: ... though as above, it's possible (and probably sensible in the case of a legacy codebase) to initially only apply gates to new code, so this doesn't have to trigger a big-bang activity.

File: quality-checks.md [hunk 10]
```diff
@@ -45,17 +45,17 @@ We recommend tracking progress on an Engineering Quality dashboard, for example:
...
+| Code smells scan       | Maintainability | Universal     | Check whether the software code adheres to the principles, patterns and practices of writing clean code                                                                                                         | The code is written once but read and executed many times. If the code is not clean, the cost and risk of making software changes both increase exponentially over time | Must use SonarQube's default rules, profiles and gateways <br/><br/> Build pipeline must fail if gateway is not met | [SonarQube](tools/sonarqube.md) | [Clean Code: Smells and Heuristics](https://moderatemisbehaviour.github.io/clean-code-smells-and-heuristics/) |
```
Comment (andyblundell) [inline quality-checks.md:58]: Can we maybe include a brief "at time of writing, this currently means a score of A for new code" or similar? Just for people who don't drill into the sub-pages?

**Review by andyblundell — COMMENTED** [review]
Looks really good, thanks - a few more minor comments...

File: tools/sonarqube.md (hunk 5 above)
Comment (dlavender4) [reply tools/sonarqube.md:19]: aww worked out why all the americanisms. we pilfered this from Sonarqube's website :)

File: quality-checks.md (hunk 10 above)
Comment (dlavender4) [reply quality-checks.md:58]: there's a bit too many

image

I'll list them in the sonarqube page, and put a link here.

File: tools/sonarqube.md [hunk 11]
```diff
@@ -0,0 +1,85 @@
...
+Quality Gates enforce a quality policy in your organisation by answering one question: is my project ready for release?
+
+To answer this question, you define a set of conditions against which projects are measured. For example "No new blocker issues", "Code coverage on new code greater than X%".
+
+We recommend that you define one gate for [new code](#definition-of-new-code), and another that covers **all** code (new and legacy).
+This approach facilitates getting new code right, without having to instantly fix legacy code.
+
+## Definition of new code
+New code is defined by SonarQube as:
+The entire contents of any file that was changed as part of this commit.
```
Comment (iw-ezequiel) [inline tools/sonarqube.md:39]: Is new code not just the changes over a given period of time, rather than the whole file where changes are present?

**Review by stefaniuk — APPROVED** [review]
LGTM 👍

**Comment by andyblundell** [comment]
This is looking really good. We have an outstanding question over the precise definition of "new code" that we want to use - I suggest we merge this in as we have at least one team actively debating how best to address a legacy codebase, so it would be helpful to get this merged ASAP. We can continue to iterate the notes.

**Review by striggy77 — APPROVED** [review]

**Pull request merged**